src/MonoMod.Core/Platforms/PlatformTriple.cs

```diff
@@ -626,106 +626,234 @@ public MethodBase GetRealDetourTarget(MethodBase from, MethodBase to)
+            for (var i = 0; i < arguments.Length; i++)
+            {
+                if (arguments[i] is SpecialArgumentKind.GenericContext)
+                    return true;
             }
+            return false;
+        }
+
+        private static bool RequiresGenericContext(MethodBase method)
+        {
+            // If neither the method nor its declaring type is generic,
+            // we do not need to worry about the generic context at all.
+            var declaringType = method.DeclaringType ?? typeof(object);
+            if (!method.IsGenericMethod && !declaringType.IsGenericType)
+                return false;
+
+            // If the method itself is generic, check whether at least one
+            // of its generic arguments makes it a shared instantiation.
+            var methodGenericArguments = method.IsGenericMethod ? method.GetGenericArguments() : Type.EmptyTypes;
+            if (methodGenericArguments.Any(static x => !x.IsValueType))
+                return true;
 
-            return to;
+            // If the method is effectively generic (i.e., it is defined on a generic type),
+            // we need to determine whether it still requires a hidden argument in the form
+            // of a MethodDesc, MethodTable, or TypeHandle pointer, or if `this` alone will
+            // suffice for this purpose.
+            //
+            // The rules are as follows:
+            //  - If both the method and its declaring type are generic, `this` alone cannot
+            //    provide the generic context the VM needs. In such cases,
+            //    we require a hidden argument for shared instantiations.
+            //  - If the method is static, there is obviously no `this` from which
+            //    the VM could infer the generic context.
+            //  - If `this` is a value type, the unboxed `this` pointer, by definition,
+            //    does not have an associated MethodTable pointer.
+            //  - If the method is a default interface method called via interface dispatch,
+            //    the VM cannot use `this` to derive the generic context because it is too
+            //    ambiguous (e.g., it may implement multiple IFoo<T> interfaces).
+            //
+            // See:
+            // https://github.com/dotnet/runtime/blob/55eee324653e01cf28809d02b25a5b0894b58d22/src/coreclr/vm/method.cpp#L1649
+            var mayNeedHiddenArg = method.IsGenericMethod
+                || method.IsStatic
+                || declaringType.IsValueType
+                || declaringType.IsInterface && !method.IsAbstract;
+            if (!mayNeedHiddenArg)
+                return false;
+
+            // Finally, check whether the type on which the method is defined may be shared.
+            var typeGenericArguments = declaringType.IsGenericType ? declaringType.GetGenericArguments() : Type.EmptyTypes;
+            return typeGenericArguments.Any(static x => !x.IsValueType);
```

Comment:
All of this logic is runtime-specific, and so should call some runtime-specific helper to make this determination. The rules implemented here are also wrong: it mishandles the case of `M<SomeStruct<string>>()` *which is still shared*, even though the type argument is a value-type. (I'll point you over [here](https://github.com/nike4613/MonoMod/tree/coreclr-generics) and [here](https://github.com/MonoMod/MonoMod.Common/tree/15d444c207a49232318f423392177f0dbea82548/RuntimeDetour/Platforms/Generic) for something to reference-- I'm pretty sure I got the logic fully correct there.)

Reply:
> All of this logic is runtime-specific

Yup, I initially wanted to put it in `IRuntime`, but that would be a breaking change since it impacts a public API. So, instead, I decided to place the method here to avoid a more significant change, as it realistically only affects CoreCLR at the moment :)

If you're okay with `IRuntime.RequiresGenericContext(MethodBase)` *(or something like that)*, I'll be happy to fix it and move the method to `Core[...]Runtime` while creating stubs for the other runtimes.  

> it mishandles the case of `M<SomeStruct<string>>()`

Oh, right - great catch! I basically followed the same CLR logic you referenced in the comments of your code but made the mistake of trivializing shared generic types to just non-value types, while somehow forgetting in the moment that value types can also be generic and, thus, shared :D

Reply:
The way I'm treating Core versioning is that any non-*consumer* breaking change is permitted across minor version bumps, and that certain consumer-breaking changes are permitted as long as they're not at the main entrypoint (i.e. DetourFactory). If we really don't like adding a method to `IRuntime`, then we can also add another interface to the implementers of `IRuntime`, like I did for CET, but I'd prefer just changing `IRuntime`.